source/Target/SwiftLanguageRuntime.cpp

```diff
@@ -2479,59 +2418,89 @@ bool SwiftLanguageRuntime::GetDynamicTypeAndAddress(
+  // Type kinds with metadata don't need archetype binding.
   if (is_indirect_enum_case)
+    // ..._IndirectEnumCase() recurses, no need to bind archetypes.
     success = GetDynamicTypeAndAddress_IndirectEnumCase(
         in_value, use_dynamic, class_type_or_name, address);
+  else if (type_info.AnySet(eTypeIsClass) ||
+           type_info.AllSet(eTypeIsBuiltIn | eTypeIsPointer | eTypeHasValue))
+    success = GetDynamicTypeAndAddress_Class(
+        in_value, *scratch_ctx, use_dynamic, class_type_or_name, address);
+  else if (type_info.AnySet(eTypeIsProtocol))
+    success = GetDynamicTypeAndAddress_Protocol(
+        in_value, *scratch_ctx, use_dynamic, class_type_or_name, address);
+  else if (type_info.AnySet(eTypeIsGenericTypeParam))
+    // ..._GenericTypeParam performs the archetype binding *and* sets address.
+    success = GetDynamicTypeAndAddress_GenericTypeParam(
+        in_value, *scratch_ctx, use_dynamic, class_type_or_name, address);
+  else if (type_info.AnySet(eTypeIsTuple))
+    // GetDynamicTypeAndAddress_Tuple recursively iterates over all children.
+    success = GetDynamicTypeAndAddress_Tuple(
+        in_value, *scratch_ctx, use_dynamic, class_type_or_name, address);
   else {
-    CompilerType var_type(in_value.GetCompilerType());
-    Flags type_info(var_type.GetTypeInfo());
-    if (type_info.AnySet(eTypeIsSwift)) {
-      if (type_info.AnySet(eTypeIsClass))
-        success = GetDynamicTypeAndAddress_Class(in_value, use_dynamic,
-                                                 class_type_or_name, address);
-      else if (type_info.AnySet(eTypeIsEnumeration))
-        success = GetDynamicTypeAndAddress_Enum(in_value, use_dynamic,
-                                                class_type_or_name, address);
-      else if (type_info.AnySet(eTypeIsProtocol))
-        success = GetDynamicTypeAndAddress_Protocol(
-            in_value, use_dynamic, class_type_or_name, address);
-      else if (type_info.AnySet(eTypeIsGenericTypeParam))
-        success = GetDynamicTypeAndAddress_GenericTypeParam(
-            in_value, use_dynamic, class_type_or_name, address);
-      else if (type_info.AnySet(eTypeIsTuple))
-        success = GetDynamicTypeAndAddress_Tuple(in_value, use_dynamic,
-                                                 class_type_or_name, address);
-      else if (type_info.AnySet(eTypeIsStructUnion))
-        success = GetDynamicTypeAndAddress_Struct(in_value, use_dynamic,
-                                                  class_type_or_name, address);
-      else if (type_info.AllSet(eTypeIsBuiltIn | eTypeIsPointer |
-                                eTypeHasValue))
-        success = GetDynamicTypeAndAddress_Class(in_value, use_dynamic,
-                                                 class_type_or_name, address);
-    }
+    // Perform archetype binding in the scratch context.
+    auto *frame = in_value.GetExecutionContextRef().GetFrameSP().get();
+    if (!frame)
```

Comment:
We always have a frame here, I think.

Reply:
Why? Couldn't in_value be an out-of-date value object whose frame doesn't exist any more?

Reply:
My understanding is that we never get to dynamic type resolution if our frame is empty (there are several layers of `if (!frame) return false`, IIRC, but I can't (as well!) prove all the paths converging here have the check.